SAMPLES, = glob_wildcards("../data/fastq/{sample}_pass.fastq")

REF = "../ref/GCF_000002195.4_Amel_4.5_genomic.fna"
GFF = "../ref/GCF_000002195.4_Amel_4.5_genomic.gff"

THREADS = 12

rule all:
	input: expand("../temp/alignments/{sample}_RG.bam",sample=SAMPLES)

rule buildBowtie:
	input: REF
	output: "../ref/Amel.1.bt2"
	shell: "bowtie2-build {REF} ../ref/Amel"

rule alignReads:
	input: "../ref/Amel.1.bt2",samp="../data/fastq/{sample}_pass.fastq"
	output: "../temp/alignments/{sample}.bam"
	shell: """bowtie2 -x ../ref/Amel -U {input[samp]} -p {THREADS} | samtools view -bSF4 > {output}; rm {input[samp]}"""

#Sorts the bam files; necessary for at least samtools
rule PicardGroups:
	input: "../temp/alignments/{sample}.bam"
	output: "../temp/alignments/{sample}_RG.bam"
	shell: """java -jar {PICARD} AddOrReplaceReadGroups I={input} O={output} SORT_ORDER=coordinate \
				CREATE_INDEX=true RGPL=illumina RGID=11 RGSM=mysample RGLB=lib1 RGPU=unit1"""